### PR TITLE
feat(timeline): add pipeline and UI for execution steps and variable diff

### LINKED ISSUE #10: Issue 1: Timeline UI + Data Pipeline
## Issue 1: Timeline UI + Data Pipeline

**Feature:** Visual Variable Timeline

### Goal
Build the Timeline UI panel and the pipeline that turns tracer events into UI-ready steps. Deliver an end-to-end flow that runs outside IDLE first.

### Scope
- **`Lib/idlelib/timeline_ui.py`** (new): Tkinter panel with:
  - Left: step list (Listbox or Treeview)
  - Right: details pane (Text or Treeview) for locals snapshot + diff summary
  - API: `set_steps(steps)`, `select_step(i)`
- **`Lib/idlelib/timeline_pipeline.py`** (new): Read tracer events, feed into TimelineStore, produce steps with schema: `index`, `filename`, `lineno`, `funcname`, `locals`, `diff`
- **`demo_timeline_ui.py`** (new): Standalone harness that runs tracer → pipeline → UI with a small target function
- Add filtering (e.g. ignore `__builtins__`) in pipeline

### Acceptance Criteria
- [ ] `demo_timeline_ui.py` runs end-to-end (trace → steps → UI)
- [ ] Step list and details pane update when selecting steps
- [ ] Step schema matches: `index`, `filename`, `lineno`, `funcname`, `locals`, `diff`
- [ ] No dependency on IDLE integration yet

### Notes
- Keep UI module independent (pure Tkinter)
- Outputs must be pickleable (plain dicts/strings)
- Default diff scope: locals only

### PR DESCRIPTION
**Pull request by evanedreo**
Closes #10

## Summary

This PR implements **Issue 1: Timeline UI + Data Pipeline** for the Visual Variable Timeline feature.

It delivers:

* A UI-agnostic pipeline that converts raw tracer events into UI-ready step dictionaries
* A standalone Tkinter Timeline panel for visualizing execution steps
* A full end-to-end demo script: tracer → pipeline → UI
* No dependency on IDLE integration yet

The feature can now be verified independently before wiring into IDLE.

---

### 1. `Lib/idlelib/timeline_pipeline.py`

Adds `events_to_steps(events)`:

* Consumes tracer events from `timeline_tracer.get_events()`
* Feeds events through `TimelineStore`
* Produces plain, pickleable step dicts with schema:

```python
{
    "index": int,
    "filename": str,
    "lineno": int,
    "funcname": str,
    "locals": dict[str, str],
    "diff": {
        "added": dict[str, str],
        "removed": dict[str, str],
        "changed": dict[str, {"before": str, "after": str}],
    },
}
```

Key behavior:

* Filters noisy keys such as `__builtins__`
* Safely stringifies values via `_safe_repr`
* Computes diffs between consecutive locals snapshots
* Ensures output is plain dicts and strings, fully pickleable

---

### 2. `Lib/idlelib/timeline_ui.py`

Reworked TimelinePanel as a pure Tkinter `Frame`:

Layout:

* Left panel: execution step list
* Right panel: locals snapshot + diff summary

Public API:

* `set_steps(steps)`
  Replaces internal state and populates step list
* `select_step(i)`
  Selects a step and updates details pane

UI behavior:

* Selecting a step updates header and locals view
* Diff section shows Added, Removed, and Changed variables
* Gracefully handles empty locals or no changes
* Optional `on_select` callback supported

The module remains independent of IDLE integration.

---

### 3. `demo_timeline_ui.py`

Standalone end-to-end harness:

1. Starts tracing
2. Runs a small `target()` function
3. Stops tracing
4. Converts events via `events_to_steps()`
5. Launches TimelinePanel
6. Calls `set_steps(steps)`

This validates the full flow outside IDLE.

Run with:

```bash
./python demo_timeline_ui.py
```

---

## Acceptance Criteria Check

* Trace → pipeline → UI works end-to-end
* Step list updates correctly
* Selecting steps updates details pane
* Step schema matches required structure
* UI module is pure Tkinter
* Outputs are pickleable
* No IDLE integration yet

All criteria satisfied.

---

## Notes

* Default diff scope is locals only
* Snapshot sanitization ensures stable, display-safe values
* Designed for clean integration in Issue 2

### CODE REVIEW

**Comment by evanedreo** [comment]
Sorry I accidentally put codex review, I won't use it

File: Lib/idlelib/timeline_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+except Exception:
+    try:
+        # When imported via a `Lib` root on sys.path.
+        from idlelib.timeline_store import TimelineStore, diff
+    except Exception:
+        # When running standalone with only `Lib/idlelib` on sys.path.
+        from timeline_store import TimelineStore, diff  # type: ignore
+
+
+_NOISY_KEYS = {"__builtins__"}
```
Comment (NicholasTc) [inline Lib/idlelib/timeline_pipeline.py:25]: Good call filtering these out — it keeps the locals view clean and beginner-friendly. 
One thing to consider for later: making this list configurable
so users can customize what gets filtered. Not needed now, just worth keeping in mind!

File: Lib/idlelib/timeline_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+
+def _safe_repr(value: Any) -> str:
```
Comment (NicholasTc) [inline Lib/idlelib/timeline_pipeline.py:28]: Nice defensive implementation! One small note: this function looks similar to 
`_safe_repr` in timeline_tracer.py and timeline_store.py. It works perfectly as-is, 
but it might be worth extracting it into a shared utility module down the line to 
avoid maintaining the same logic in multiple places. Great job keeping it safe and crash-proof though!

File: demo_timeline_ui.py [hunk 3]
```diff
@@ -19,126 +25,56 @@
...
+    # running interpreter version). We only add `Lib/idlelib` so we can import
+    # the timeline modules without shadowing unrelated stdlib modules.
     import sys
-    if './Lib' not in sys.path:
-        sys.path.insert(0, './Lib')
-    
-    from idlelib.timeline_ui import TimelinePanel
-    
-    # Create hidden root window; panel uses Toplevel
+    idlelib_dir = "./Lib/idlelib"
```
Comment (katew-23) [inline demo_timeline_ui.py:43]: Nice call avoiding full Lib/ insertion. One suggestion: this relative path assumes the script is run from repo root. Would you consider deriving it from Path(__file__).resolve().parent so it works regardless of current working directory? Or maybe state somewhere that this should be run from the repo root.

File: Lib/idlelib/timeline_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    try:
+        return repr(value)
+    except Exception as exc:
+        return f"<unreprable {type(value).__name__}: {exc!r}>"
+
+
+def _sanitize_locals_snapshot(snapshot: Any) -> dict[str, str]:
```
Comment (katew-23) [inline Lib/idlelib/timeline_pipeline.py:35]: This sanitization helper is clean and easy to reason about. Filtering noisy keys plus defensive repr handling is a good touch for UI stability.

**Comment by katew-23** [comment]
# Peer Code Review

## Summary
This PR adds the first end-to-end timeline feature slice outside IDLE integration: a pipeline that converts tracer events into UI-ready step dicts, a standalone Tkinter timeline panel, and a demo harness that runs tracer → pipeline → UI. The code is generally modular and readable, with clear separation between data transformation and rendering.

---

## Correctness
Check the box for each requirement you can confirm is satisfied based on the PR and evidence (tests, screenshots, docs, output, reasoning from code):

- [x] Core functionality works as intended
- [x] Handles edge cases and error conditions appropriately

### Evidence / Notes
- I reviewed the implementation in `Lib/idlelib/timeline_pipeline.py`, `Lib/idlelib/timeline_ui.py`, and `demo_timeline_ui.py`.
- I verified syntax/compile validity with:
  - `./python.exe -m py_compile Lib/idlelib/timeline_pipeline.py Lib/idlelib/timeline_ui.py demo_timeline_ui.py`
- Existing related tests pass:
  - `PYTHONPATH=./Lib ./python.exe -m unittest -v idlelib.idle_test.test_timeline_store idlelib.idle_test.test_timeline_tracer`
- Potential correctness/maintainability concerns:
  - Demo path insertion uses a cwd-dependent relative path (`./Lib/idlelib`), which may fail when run outside repo root.

---

## Design 
Choose one:

- [x] Exemplary - Clean, modular, easy to follow. Strong separation of concerns. Minimal complexity.
- [ ] Satisfactory - Works well overall with minor complexity/redundancy or small opportunities to refactor.
- [ ] Needs Improvement - Hard to follow, overly complex, tightly coupled, or missing structure.
- [ ] Missing/None - Design is unclear or not reasonably maintainable.

### Explain your choice
- The PR has good separation of concerns:
  - `timeline_pipeline.py` handles event normalization/diff preparation.
  - `timeline_ui.py` focuses on rendering and user interaction.
  - `demo_timeline_ui.py` wires up tracing and display for independent verification.
- Overall structure is easy to follow and positioned well for later IDLE integration.

---

## Code Style
Choose one:

- [x] Exemplary - Consistent formatting, clear naming, helpful comments, follows style guide.
- [ ] Satisfactory - Minor style inconsistencies, but overall readable.
- [ ] Needs Improvement - Naming/formatting/commenting issues reduce readability.
- [ ] Missing/None - No consistent style; readability is severely impacted.

### Explain your choice
- [x] Variable/function/class names are descriptive and consistent
- [x] Formatting/indentation is consistent
- [x] Comments/docstrings clarify non-obvious logic
- [x] Code follows project/course style guidelines

## Inline Comments 
Did you add any inline comments in the Files Changed tab? (So that the author of the PR doesn’t miss your inline comment)

- [x] Yes
- [ ] No

**Comment by evanedreo** [comment]
closed because already included in #24  (Nicholas used my branch to implement another feature then merged it, so we don't need to merge mine)
